[50254] Filtering on share modal

HDinger opened the pull request:
- [x] (Frontend) Add filter option for type and role in the box header
- [x] (Frontend) Toggle with the additional bulkActions when selecting any checkbox
- [x] (Frontend) Show currently active filter in the dropdown
- [x] (Backend) Actually filter when selecting a filter
  - [x] Extend MemberQuery by a filter for the entity
  - [x] Filter for Role
  - [x] Filter for Type
- [x] Add Test cases for filtering

**Open Questions**
- [x] Unset a filter
- [x] Keep role filter when selecting a type filter (and the other way around)

**Handled separatly**
- [x] ~When the filter results are empty, the Box switches to the blankslate, where no filter is available. Thus I cannot leave this stage any more.~ See [ticket](https://community.openproject.org/projects/openproject/work_packages/50931/activity)
- [x] ~Inherited permissions are in filter result set but might show a different permission if the user was shared separatly~ See [ticket](https://community.openproject.org/projects/openproject/work_packages/50939/activity) 

https://community.openproject.org/projects/openproject/work_packages/50254/activity

HDinger committed:
Add (dummy) action menus for filtering the list of shares in the modal

HDinger committed:
Toggle between filters and bulkActions in the share modal

HDinger committed:
Add test cases for filtering the shareWP modal

HDinger committed:
Extend MemberQuery by filters for the entity and use that in the shares_controller

HDinger committed:
Add role filter for share modal

HDinger committed:
Replace form by direct links as it enables us to reset a filter without any additional js required

HDinger committed:
Add (empty) type filter options to the share modal

HDinger committed:
Add principal type filter for the sharing modal

HDinger committed:
Add additional test case for using both filters at the same time

klaustopher committed:
Add AlsoProjectMemberFilter for members

HDinger committed:
Add "AND" keyword to sql query && always set the entity to the current WP

HDinger committed:
Add check for currenltly active type filter

HDinger commented:
Hi @aaron-contreras 

I added the still open points to the description. Futher, the code needs some cleanup, especially [this part](https://github.com/opf/openproject/pull/14075/files#diff-dcc2c760c3da0ae92ef0691162adad19cc81662a330b2a062405da3c68fa139fR92-R143). Other than that this ready for a first review.

aaron-contreras commented on app/components/work_packages/share/bulk_permission_button_component.html.erb:
```diff
@@ -4,7 +4,7 @@
                                        color: :subtle,
                                        data: { test_selector: 'op-share-wp-bulk-update-role'})) do |menu|
-    menu.with_show_button(data: { 'work-packages--share--bulk-selection-target': 'bulkUpdateRoleLabel' }) do |button|
+    menu.with_show_button(scheme: :invisible, color: :subtle, data: { 'work-packages--share--bulk-selection-target': 'bulkUpdateRoleLabel' }) do |button|
```
I fiddled around a bunch trying to find the scheme that would make the buttons look like they do with this PR. Never thought of combining a scheme and a color. Thank you :)

aaron-contreras commented on app/components/work_packages/share/modal_body_component.rb:
```diff
@@ -30,14 +30,17 @@ module WorkPackages
         super
 
         @work_package = work_package
+        @shared_principals = shares
```
I'd rename this to `@share` since now it no longer houses a `Principal` Relation but instead a `Member` relation

HDinger replied:
mhh `@share` has led to some issues when inline creating as the same name is used there. I changed it now to `@shared_members` 🤷‍♀️

aaron-contreras commented on frontend/src/stimulus/controllers/dynamic/work-packages/share/bulk-selection.controller.ts:
```diff
@@ -126,9 +128,11 @@ export default class BulkSelectionController extends Controller {
     this.toggleAllTarget.checked = checkedSharesCount === sharesCount;
 
     if (this.checked.length === 0) {
-      this.actionsTarget.setAttribute('hidden', 'true');
+      this.bulkActionsTarget.setAttribute('hidden', 'true');
+      this.defaultActionsTarget.removeAttribute('hidden');
     } else {
-      this.actionsTarget.removeAttribute('hidden');
+      this.bulkActionsTarget.removeAttribute('hidden');
+      this.defaultActionsTarget.setAttribute('hidden', 'true');
       this.updateBulkUpdateRoleLabelValue();
```
👍

aaron-contreras reviewed (commented):
Very nice so far, besides the open points and known issues you mentioned, it looks good to me. I'd like to see what the implementation turns out to be once we can chain both filter types at a time. Great job so far!

HDinger committed:
Kepp currently selected role filter when filtering the type (and the other way around). Further, the test has been adapted to it.

HDinger committed:
Adapt tests to new members filters

HDinger commented:
Hi @aaron-contreras 

I excluded the inheritance problem to [#50939](https://community.openproject.org/projects/openproject/work_packages/50939/activity) as I am unsure about the expected results and would like to discuss that with product first.

aaron-contreras committed:
Merge branch 'dev' into implementation/50254-filtering-on-share-modal

aaron-contreras committed:
Fix: Register `WorkPackageMemberQuery`

The `MemberQuery` has a default_scope that checks a non-admin has a
certain set of permissions which are not required for the behavior of
the share modal (or visibility of work package shares for that matter).

Given that visibility is unscoped for the share modal, extracting a
separate query class that inherits from the MemberQuery sans the
default_scope made sense as a way to keep these two in sync minus
the varying visibility levels based on the current_user.

aaron-contreras committed:
Fix: Ammend stream actions to enable functional responsiveness behavior

There were a couple of specs failing due to the new query layer and
injectable shares into the ModalBodyComponent. This commit standardizes
the interface to prevent this in an easy to follow manner.

aaron-contreras committed:
Cleanup: Remove extracted border box custom patches

Thanks for extracting Attila!

aaron-contreras reviewed (approved):
Went ahead and fixed the pesky issues you were facing. You can merge whenever you're back :)

HDinger commented:
Thanks a lot, @aaron-contreras for jumping in! 🤗
